File: solutions/724. Find Pivot Index.cpp
class Solution {
public:
    int pivotIndex(vector<int>& nums) {
        /* One may use `std::accumulate` to calculate the sum concisely
            but that puts extra overhead. */
        int sum = 0;
        for (const int num : nums) { sum += num; }
        /* Since we're looking for the left-most pivot,
            a window slides from the left to the right,
            dynamically checking the sum of both sides. */
        int rightSum = sum;
        int leftSum = 0;
        for (size_t i = 0; i < nums.size(); ++i) {
            if (leftSum == rightSum - nums.at(i)) {
                return i;
            }
            rightSum -= nums.at(i);
            leftSum += nums.at(i);
        }
        return -1;
    }
};

File: solutions/724. Find Pivot Index.rs
impl Solution {
    pub fn pivot_index(nums: Vec<i32>) -> i32 {
        let prefix_sum = nums
            .iter()
            .scan(0, |state, &num| {
                *state += num;
                Some(*state)
            })
            .collect::<Vec<_>>();
        let sum = prefix_sum[nums.len() - 1];
        for i in 0..nums.len() {
            let left_sum = sum - prefix_sum[i];
            let right_sum = if i == 0 { 0 } else { prefix_sum[i - 1] };
            if left_sum == right_sum {
                return i as i32;
            }
        }
        -1
    }
}
